src/components/shared/qrcode.tsx

```diff
@@ -16,26 +25,19 @@ const cache = new Proxy(sessionStorage, {
     },
 })
 
-export function QRCode(props: {
-    text: string
-    options?: qr.QRCodeRenderersOptions
-    canvasProps?: React.DetailedHTMLProps<
-        React.CanvasHTMLAttributes<HTMLCanvasElement & HTMLImageElement>,
-        HTMLCanvasElement & HTMLImageElement
-    >
-}) {
+export function QRCode({ text, options = {}, canvasProps }: QRProps) {
     const ref = useRef<HTMLCanvasElement | null>(null)
-    const image = cache.get(props.text)
+    const image = cache.get(text)
     useEffect(() => {
-        if (cache.get(props.text) || !ref.current) return
-        qr.toCanvas(ref.current, props.text, props.options)
-        return () => {
-            // if already rendered canvas, do not re-render img
-            // eslint-disable-next-line react-hooks/exhaustive-deps
-            cache.set(props.text, ref.current?.toDataURL())
-        }
-    }, [props.options, props.text])
-    return image ? <img src={image} {...props.canvasProps} /> : <canvas {...props.canvasProps} ref={ref} />
+        if (!ref.current) return
```

Comment:
Although `useEffect` runs after every render cycle, which means we can get ref first time. But It is the case that `ref` must be `undefined` since it also returns <img />. So `if (!ref.current) return` is still necessary .